models/avatars/avatar.go

```diff
@@ -110,39 +104,73 @@ func HashedAvatarLink(email string, size int) string {
+}
+
+// GenerateUserAvatarFastLink returns a fast link (302) to the user's avatar: "/user/avatar/${User.Name}/${size}"
+func GenerateUserAvatarFastLink(userName string, size int) string {
+	if size < 0 {
+		size = 0
 	}
-	return setting.AppSubURL + "/avatar/" + url.PathEscape(sum)
+	//AppSubURL: It is either "" or starts with '/' and ends without '/', such as '/{subpath}'. question: should we use AppURL or StaticURLPrefix?
+	return setting.AppSubURL + "/user/avatar/" + userName + "/" + strconv.Itoa(size)
 }
 
-// MakeFinalAvatarURL constructs the final avatar URL string
-func MakeFinalAvatarURL(u *url.URL, size int) string {
-	vals := u.Query()
-	vals.Set("d", "identicon")
-	if size != DefaultAvatarSize {
-		vals.Set("s", strconv.Itoa(size))
+// GenerateUserAvatarImageLink returns a link for `User.Avatar` image file: "/avatars/${User.Avatar}"
+func GenerateUserAvatarImageLink(userAvatar string, size int) string {
+	if size > 0 {
+		return setting.AppSubURL + "/avatars/" + userAvatar + "?size=" + strconv.Itoa(size)
 	}
-	u.RawQuery = vals.Encode()
-	return u.String()
+	return setting.AppSubURL + "/avatars/" + userAvatar
 }
 
-// SizedAvatarLink returns a sized link to the avatar for the given email address.
-func SizedAvatarLink(email string, size int) string {
+// generateEmailAvatarLink returns a email avatar link.
+// if final is true, it may use a slow path (eg: query DNS).
+// if final is false, it always uses a fast path.
+func generateEmailAvatarLink(email string, size int, final bool) string {
+	email = strings.TrimSpace(email)
+	if email == "" {
+		return DefaultAvatarLink()
+	}
+
 	var avatarURL *url.URL
+	var err error
+
 	if setting.EnableFederatedAvatar && setting.LibravatarService != nil {
-		// This is the slow path that would need to call LibravatarURL() which
-		// does DNS lookups. Avoid it by issuing a redirect so we don't block
-		// the template render with network requests.
-		return HashedAvatarLink(email, size)
+		emailHash := saveEmailHash(email)
+		if final {
+			if avatarURL, err = LibravatarURL(email); err != nil {
+				return DefaultAvatarLink()
+			}
+		} else {
+			if size > 0 {
+				return setting.AppSubURL + "/avatar/" + emailHash + "?size=" + strconv.Itoa(size)
+			}
+			return setting.AppSubURL + "/avatar/" + emailHash
+		}
 	} else if !setting.DisableGravatar {
-		// copy GravatarSourceURL, because we will modify its Path.
-		copyOfGravatarSourceURL := *setting.GravatarSourceURL
-		avatarURL = &copyOfGravatarSourceURL
+		avatarURLDummy := *setting.GravatarSourceURL // copy GravatarSourceURL, because we will modify its Path.
+		avatarURL = &avatarURLDummy
 		avatarURL.Path = path.Join(avatarURL.Path, HashEmail(email))
 	} else {
 		return DefaultAvatarLink()
 	}
 
-	return MakeFinalAvatarURL(avatarURL, size)
+	urlQuery := avatarURL.Query()
+	urlQuery.Set("d", "identicon")
+	if size > 0 {
+		urlQuery.Set("s", strconv.Itoa(size))
+	}
+	avatarURL.RawQuery = urlQuery.Encode()
+	return avatarURL.String()
```

Comment:
might be better to extract this out to a function then you can return the result of this function at 144 and at 154.

That would allow you to simplify some of the logic above.

Reply:
But there is no caller to these codes now. It only works for gravatar-like URLs. The line 114 doesn't share the same logic as here.

I think we can postpone the function extraction to next time when others need it?

Reply:
Oh, I know your meaning, just a minute ~~~